# Single-Cell RNA Sequencing (scRNA-seq) Analysis

## Tutorials
1. [Datacamp](https://campus.datacamp.com/courses/single-cell-rna-seq-workflows-in-r/)
2. [Hemberg's Lab course](https://scrnaseq-course.cog.sanger.ac.uk/website/index.html)
3. [OCW MIT Computational Biology](https://ocw.mit.edu/courses/electrical-engineering-and-computer-science/6-047-computational-biology-fall-2015/readings/)
4. [Computational Biology: Genomes, Networks, Evolution*](http://ocw.mit.edu/ans7870/6/6.047/f15/MIT6_047F15_Compiled.pdf)
5. [MIT's 6.047/6.878  Machine Learning in Genomics*](https://stellar.mit.edu/S/course/6/fa19/6.047/materials.html)
6. [Deep Learning for Single Cell Biology](https://towardsdatascience.com/deep-learning-for-single-cell-biology-935d45064438)
7. [Seurat - Guided Clustering Tutorial*](https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html)

## Important GitHub profiles and repositories
1. [hemberg-lab](https://github.com/hemberg-lab)
2. [SingleCellThoughts](https://github.com/LTLA/SingleCellThoughts), [SingleCellThoughts](https://ltla.github.io/SingleCellThoughts/)
3. [Single-cell exploratory data analysis for RNA-Seq](https://github.com/logstar/scedar)
4. [Single-cell RNA sequencing (Cell Ranger)](https://bioinformatics.uconn.edu/single-cell-rna-sequencing-cell-ranger-2/#)
5. [Nikolay Oskolkov](https://github.com/NikolayOskolkov)
6. [Awesome Single Cell](https://github.com/seandavi/awesome-single-cell)
7. [scScope](https://github.com/AltschulerWu-Lab/scScope)
8. [Dendrosplit](https://github.com/jessemzhang/dendrosplit)
9. [Clustering cells from single cell RNA seq assays](https://github.com/govinda-kamath/clustering_on_transcript_compatibility_counts)

## Notable Labs' website
1. [Gilad's Lab](https://giladlab.uchicago.edu/)
2. [Kellis Lab at MIT*](http://compbio.mit.edu/)
3. [Broad Institute](https://www.broadinstitute.org/)
4. [Marioni group*](https://www.ebi.ac.uk/research/marioni)

## Resource Portals
1. [SIB ExPASy Bioformatics Resources Portal](https://www.expasy.org/)

## Useful R-Packages
1. scRNAseq
2. scater
3. scran
4. seurat
5. MAST
6. NMF

## Python Packages
1. [Biopython](https://biopython.org/)
2. [Scanpy](https://scanpy.readthedocs.io/en/stable/tutorials.html)


## YouTube Videos
1. [StatQuest: A gentle introduction to RNA-seq](https://www.youtube.com/watch?v=tlf6wYJrwKY)
2. [MIT CompBio Lecture 21 - Single-Cell Genomics, Fall 2018](https://www.youtube.com/watch?v=mJKbnJO5gZk)
3. [MIT CompBio Lecture 21 - Single-Cell Genomics, Fall 2019](https://www.youtube.com/watch?v=o7H3xNKDUig)

## Relevant Research Papers
1. [Deconstructing Olfactory Stem Cell Trajectories at Single Cell Resolution*](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5484588/pdf/nihms868754.pdf)
2. [DeepCpG: accurate prediction of single-cell DNA methylation states using deep learning*](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1189-z)
3. [Deciphering epigenomic code for cell differentiation using deep learning*](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-019-6072-8)
4. [Current best practices in single-cell RNA-seq analysis: a tutorial*](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6582955/pdf/MSB-15-e8746.pdf)
5. [Revealing the vectors of cellular identity with single-cell genomics*](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5465644/pdf/nihms857730.pdf)
6. [Benchmarking single cell RNA-sequencing analysis pipelines using mixture control experiments](https://www.nature.com/articles/s41592-019-0425-8)
7. [Single-cell RNA-seq denoising using a deep count autoencoder*](https://www.nature.com/articles/s41467-018-07931-2.pdf)
8. [12 Grand Challenges in Single-Cell Data Science*](https://peerj.com/preprints/27885.pdf)
9. [An interpretable framework for clustering single-cell RNA-Seq datasets](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2092-7)

## Theses and Dissertations
1. [Reinthinking Single-Cell RNA-Seq Analysis*](https://stacks.stanford.edu/file/druid:jr605ws6765/Stanford_PhD_Thesis-augmented.pdf)

## Datasets




---

<sup>*[web.archive.org](web.archive.org) version backup available.</sup>

## Curator
- Rahul Bhadani ( rahulbhadani@email.arizona.edu)
